src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -16,124 +16,204 @@
+            pet.setBirthDate((new SimpleDateFormat("yyyy-MM-dd")).parse(birthday));
+        } catch (ParseException e) {
+            throw new Exception("The birthday data " + birthday + " is not valid!", e);
+        }
+
+        // pet type in field 2 ["dog", "cat", ...]
+        final String petTypeString = csvRow.get(2);
+        PetType petType = petTypeMap.get(petTypeString);
+        if (petType != null) {
+            pet.setType(petType);
+        } else {
+            throw new Exception("The pet type " + petTypeString + " was not defined in the clinic pet types.");
+        }
+
+        // pet owner in field 3
+        final String owner = csvRow.get(3);
+        List<Owner> matchingOwners = petOwners.stream().filter(o -> o.getLastName().equals(owner)).collect(Collectors.toList());
+        if (matchingOwners.size() == 0) {
+            throw new Exception("Owner not found: " + owner);
+        } else if (matchingOwners.size() > 1) {
+            throw new Exception("Owner not unique: " + owner);
+        }
+        pet.setOwner(matchingOwners.iterator().next());
+
+        // pet clinic action in field 4 ("add", "delete")
+        final String clinicPetAction = csvRow.get(4);
+        switch (clinicPetAction.toLowerCase()) {
+            case "add":
+            case "":
+                return new PetRowResult(pet, ClinicPetAction.ADD);
+            case "delete":
+                return new PetRowResult(pet, ClinicPetAction.DELETE);
+            default:
+                throw new Exception("Unknown pet action: " + clinicPetAction);
+        }
+    }
+
+    /**
+     * Tests if two pets are equal. This should normally be done in the Pet class.
+     * This only tests if the name, the type and the birthday are equal.
+     *
+     * @param pet1 first pet to compare
+     * @param pet2 second pet to compare
+     * @return boolean if the pets are equal
+     */
+    public boolean arePetsEqual(Pet pet1, Pet pet2) {
+        return pet2.getName().equals(pet1.getName()) &&
+            pet2.getType().getId().equals(pet1.getType().getId()) &&
+            pet2.getBirthDate().equals(pet1.getBirthDate());
+    }
+
+    /**
+     * Removes a pet from the clinic service
+     *
+     * @param pet Pet which should be removed
+     */
+    public void removePetFromClinicService(Pet pet) {
+        for (Pet sameOwnerPet : pet.getOwner().getPets()) {
+            if (arePetsEqual(pet, sameOwnerPet)) {
+                clinicService.deletePet(sameOwnerPet);
+            }
+        }
+    }
+
     @PreAuthorize("hasRole(@roles.OWNER_ADMIN)")
     @RequestMapping(value = "importPets",
         method = RequestMethod.POST,
         consumes = "text/plain",
         produces = "application/json")
     public ResponseEntity<List<Pet>> importPets(@RequestBody String csv) {
-
-        int i = 0;
         List<Pet> pets = new LinkedList<Pet>();
-        Pet pet;
 
-        do {
-            pet = new Pet();
+        // map the csv string to a matrix
```

Comment:
I like that you kept the indentations small except maybe in this class...so you could think about also extracting some methods here for each purpose (e.g., the last switch case)